# Предсказывание стоимости акций

**Задача**: Обучить несколько моделей и выбрать лучшую, которая будет предсказывать стоимость акции открытия на следующий день.

Данная работа ещё в процессе. Ожидается добавление всех этапов, их более подробное описание, перевод на английский язык, а также новые попытки улучшения метрики.

## Загрузка данных

In [1]:
# Загрузка необходимых библиотек
import warnings
import re
import os

import pandas as pd
import numpy as np
import yfinance as yf

import seaborn as sns
import matplotlib.pyplot as plt

import optuna
from sklearn.preprocessing import StandardScaler
from sklearn.pipeline import Pipeline
from sklearn.linear_model import LinearRegression, Ridge, Lasso
from sklearn.model_selection import train_test_split, cross_val_score
from sklearn.metrics import make_scorer
from sklearn.ensemble import RandomForestRegressor, GradientBoostingRegressor
from xgboost import XGBRegressor
from sklearn.metrics import mean_squared_error

In [2]:
# настройки
warnings.filterwarnings("ignore")
optuna.logging.set_verbosity(optuna.logging.WARNING)
optuna.logging.disable_default_handler()

# константы
NAME="apple"
TARGET = "Open"
START_DATE = "2019-01-01"
END_DATE = "2025-01-01"
RANDOM_STATE = 42
TEST_SIZE = 0.3
N_TRAILS = 50

# используемые акции и их тикеты
name_of_stocks = {"apple": "AAPL",
                 "microsoft": "MSFT",
                 "nvidia": "NVDA",
                 "google": "GOOG"}

### Functions

In [3]:
# функция для просмотра данных        
def data_review(data):
    print('ПЕРВЫЕ 5 СТРОЧЕК С ДАННЫМИ:')
    display(data.head())
    print(f'\nОБЩАЯ ИНФОРМАЦИЯ О ДАННЫХ:\n')
    data.info()
    print(f'\n\nКОЛИЧЕСТВО ПРОПУСКОВ В ДАННЫХ:\n\n{data.isna().sum()}\n')
    print(f'\nКОЛИЧЕСТВО ДУБЛИКАТОВ В ДАННЫХ - {data.duplicated().sum()}\n')
    print('\nСТАТИСТИКА ПО ЧИСЛОВЫМ СТОЛБЦАМ:')
    display(data.describe())

In [4]:
def download_and_merge_stocks(stock_dict, start_date, end_date):
    """Function for downloding data and merging all list of stocks"""
    # Создаем пустой список для хранения отдельных DataFrame
    merged_data = []

    for name, ticker in stock_dict.items():
        # Загружаем данные
        stock_data = yf.download(ticker, start=start_date, end=end_date, progress=False)
        
        # Переименовываем столбцы
        stock_data = stock_data.add_suffix(f"_{name}")
        stock_data = stock_data.rename(columns={f'Adj Close_{name}': f'Adj_Close_{name}'})
        
        # Добавляем полученные данные в список
        merged_data.append(stock_data)
    
    # Объединяем все данные по индексу (дата)
    merged_df = pd.concat(merged_data, axis=1)
    
    return merged_df

### Загрузка и первичное знакомство с данными

In [5]:
data = download_and_merge_stocks(name_of_stocks, START_DATE, END_DATE)

# Просмотр первых строк объединенного DataFrame
data_review(data)

ПЕРВЫЕ 5 СТРОЧЕК С ДАННЫМИ:


,Open_apple,High_apple,Low_apple,Close_apple,Adj_Close_apple,Volume_apple,Open_microsoft,High_microsoft,Low_microsoft,Close_microsoft,...,Low_nvidia,Close_nvidia,Adj_Close_nvidia,Volume_nvidia,Open_google,High_google,Low_google,Close_google,Adj_Close_google,Volume_google
Date,,,,,,,,,,,,,,,,,,,,,
2019-01-02,38.722500,39.712502,38.557499,39.480000,37.708599,148158800,99.550003,101.750000,98.940002,101.120003,...,3.25125,3.40550,3.378090,508752000,50.828499,52.616001,50.785500,52.292500,52.105278,30652000
2019-01-03,35.994999,36.430000,35.500000,35.547501,33.952549,365248800,100.099998,100.190002,97.199997,97.400002,...,3.19225,3.19975,3.173996,705552000,52.049999,52.848999,50.703499,50.803001,50.621113,36822000
2019-01-04,36.132500,37.137501,35.950001,37.064999,35.401947,234428400,99.720001,102.510002,98.930000,101.930000,...,3.24250,3.40475,3.377346,585620000,51.629501,53.542000,51.370899,53.535500,53.343830,41878000
2019-01-07,37.174999,37.207500,36.474998,36.982498,35.323158,219111200,101.639999,103.269997,100.980003,102.059998,...,3.41075,3.58500,3.556144,709160000,53.575001,53.700001,52.737999,53.419498,53.228245,39638000
2019-01-08,37.389999,37.955002,37.130001,37.687500,35.996525,164101200,103.040001,103.970001,101.709999,102.800003,...,3.42250,3.49575,3.467614,786016000,53.805500,54.228001,53.026501,53.813999,53.621330,35298000



ОБЩАЯ ИНФОРМАЦИЯ О ДАННЫХ:

<class 'pandas.core.frame.DataFrame'>
DatetimeIndex: 1510 entries, 2019-01-02 to 2024-12-31
Data columns (total 24 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Open_apple           1510 non-null   float64
 1   High_apple           1510 non-null   float64
 2   Low_apple            1510 non-null   float64
 3   Close_apple          1510 non-null   float64
 4   Adj_Close_apple      1510 non-null   float64
 5   Volume_apple         1510 non-null   int64  
 6   Open_microsoft       1510 non-null   float64
 7   High_microsoft       1510 non-null   float64
 8   Low_microsoft        1510 non-null   float64
 9   Close_microsoft      1510 non-null   float64
 10  Adj_Close_microsoft  1510 non-null   float64
 11  Volume_microsoft     1510 non-null   int64  
 12  Open_nvidia          1510 non-null   float64
 13  High_nvidia          1510 non-null   float64
 14  Low_nvidia           1510 non-null   floa

,Open_apple,High_apple,Low_apple,Close_apple,Adj_Close_apple,Volume_apple,Open_microsoft,High_microsoft,Low_microsoft,Close_microsoft,...,Low_nvidia,Close_nvidia,Adj_Close_nvidia,Volume_nvidia,Open_google,High_google,Low_google,Close_google,Adj_Close_google,Volume_google
count,1510.000000,1510.000000,1510.000000,1510.000000,1510.000000,1.510000e+03,1510.000000,1510.000000,1510.000000,1510.000000,...,1510.000000,1510.000000,1510.000000,1.510000e+03,1510.000000,1510.000000,1510.000000,1510.000000,1510.000000,1.510000e+03
mean,136.931989,138.437162,135.555000,137.078805,135.294023,9.417009e+07,266.905715,269.488808,264.207582,266.975887,...,32.123260,32.779426,32.753858,4.484067e+08,109.724901,110.964674,108.617350,109.810141,109.449869,2.755560e+07
std,53.774234,54.217224,53.361330,53.826058,54.242113,5.233404e+07,95.597983,96.201467,94.847564,95.553237,...,36.409470,37.138628,37.143256,1.966090e+08,37.602728,37.947416,37.253635,37.581565,37.504653,1.267679e+07
min,35.994999,36.430000,35.500000,35.547501,33.952549,2.323470e+07,99.550003,100.190002,97.199997,97.400002,...,3.192250,3.199750,3.173996,9.788400e+07,50.828499,52.374500,50.676800,50.803001,50.621113,6.809800e+06
25%,91.257500,92.705626,90.653124,91.315624,88.979227,5.914420e+07,200.142502,203.724995,197.567497,201.452503,...,9.454313,9.550562,9.520722,3.093051e+08,73.536501,74.429750,72.960627,73.744753,73.480728,1.924775e+07
50%,145.540001,147.264999,144.120003,145.860001,143.596550,8.152105e+07,259.165009,261.500000,255.989998,259.464996,...,17.060250,17.606999,17.574616,4.182870e+08,110.301746,112.002750,108.640003,110.647751,110.251610,2.471900e+07
75%,173.954998,175.760002,172.285000,174.325001,172.455418,1.123920e+08,331.962494,335.080009,328.374992,331.197495,...,40.926000,41.851749,41.833623,5.449992e+08,138.356125,139.674496,137.049744,138.490120,137.994301,3.192100e+07
max,258.190002,260.100006,257.630005,259.019989,259.019989,4.265100e+08,467.000000,468.350006,464.459991,467.559998,...,146.259995,148.880005,148.869751,2.511528e+09,198.529999,202.880005,196.690002,198.160004,198.160004,1.241400e+08


Краткие пояснения, что за собой скрывают названия столбцов:

1. **Date**: Дата торгового дня.

2. **Open**: Цена открытия акций в начале торгового дня.

3. **High**: Самая высокая цена, достигнутая акциями в течение торгового дня.

4. **Low**: Самая низкая цена, достигнутая акциями в течение торгового дня.

5. **Close**: Цена закрытия акций в конце торгового дня (без учета корпоративных действий).

6. **Adj Close**: Скорректированная цена закрытия, которая учитывает все корпоративные действия, такие как выплата дивидендов, разделение акций и т. д.

7. **Volume**: Объем торгов, показывающий количество акций, которые были куплены и проданы в течение торгового дня.

## Анализ данных

### Functions

In [6]:
def new_features(df, name=NAME, sma_periods=[3, 5, 10], target=TARGET, max_lag=13):
    """
    Function is adding new features like:
    - differences between all basik columns, 
    - calculated SMA (Simple Moving Average), EMA (Exponential Moving Average) and Volatility (sma_periods)
    - Volume and Price change
    - Features of trend and seasonality
    - Lag features (max_lag)
    - Create Target (target), which is .shift(-1) of target_col
    """
    # Приведение всех столбцов к общему виду
    target_col = f'{target}_{name}'
    
    # Скользящие средние (SMA и EMA) и Волатильность
    for period in sma_periods:
        df[f'SMA_{period}_{name}'] = df[target_col].rolling(window=period).mean()
        df[f'EMA_{period}_{name}'] = df[target_col].ewm(span=period, adjust=False).mean()
        df[f'Volatility_{period}_{name}'] = df[target_col].rolling(window=period).std()
    
    # Процентные изменения
    df[f'Volume_Change_{name}'] = df[f'Volume_{name}'].pct_change()
    df[f'Price_Change_{name}'] = df[target_col].pct_change()
    
    # Создание базовых признаков (на базе разности)
    df[f'Open_High_{name}'] = df[f'Open_{name}'] - df[f'High_{name}']
    df[f'Open_Low_{name}'] = df[f'Open_{name}'] - df[f'Low_{name}']
    df[f'High_Low_{name}'] = df[f'High_{name}'] - df[f'Low_{name}']
    df[f'High_Close_{name}'] = df[f'High_{name}'] - df[f'Close_{name}']
    df[f'Low_Close_{name}'] = df[f'Low_{name}'] - df[f'Close_{name}']
    df[f'Close_Open_{name}'] = df[f'Close_{name}'] - df[f'Open_{name}']
    df[f'AdjClose_Open_{name}'] = df[f'Adj_Close_{name}'] - df[f'Open_{name}']
    df[f'AdjClose_Low_{name}'] = df[f'Adj_Close_{name}'] - df[f'Low_{name}']
    df[f'AdjClose_High_{name}'] = df[f'Adj_Close_{name}'] - df[f'High_{name}']
    df[f'AdjClose_Close_{name}'] = df[f'Adj_Close_{name}'] - df[f'Close_{name}']
    
    # Добавление признаков тренда и сезонности
    df[f'Day_of_Week_{name}'] = df.index.dayofweek
    df[f'Number_of_Week_{name}'] = df.index.isocalendar().week
    df[f'Month_{name}'] = df.index.month
    df[f'Quarter_{name}'] = df.index.quarter
    
    # Добавление лаг-функций
    for lag in range(1, max_lag):
        df[f'Lag_{lag}_{name}'] = df[target_col].shift(lag)
    
    # Добавляем столбец с ценой таргета на следующий день
    df['Target'] = df[target_col].shift(-1)
    df['Low_future'] = df[f'Low_{name}'].shift(-1)
    
    # Удаление строк с NaN значениями, которые появились из-за использования скользящих средних и других вычислений
    df.dropna(inplace=True)
    
    return df

### Feature engineering

In [7]:
data = new_features(data)

На этом этапе реализуем генерацию дополнительных признаков для обогащения данных с целью предсказания **цены открытия акций компании Apple на следующий день**. Ниже представлен подробный разбор добавленных признаков.

---

**1. Базовые признаки: разности между основными показателями**

Анализ внутридневной и междневной динамики цен. Эти признаки помогают модели учитывать волатильность и изменения цен в течение дня. Стоит также отменить, что данные признаки были созданы не только для целевой переменной, но и для других компаний.

**Признаки**:
- `Open_High`: Разница между ценой открытия (`Open`) и максимальной ценой (`High`).
- `Open_Low`: Разница между ценой открытия и минимальной ценой (`Low`).
- `High_Low`: Дневной диапазон цен (разница между `High` и `Low`).
- `High_Close`: Разница между максимальной ценой (`High`) и ценой закрытия (`Close`).
- `Low_Close`: Разница между минимальной ценой (`Low`) и ценой закрытия (`Close`).
- `Close_Open`: Разница между ценой закрытия (`Close`) и ценой открытия (`Open`).
- `AdjClose_Open`: Разница между скорректированной ценой закрытия (`Adj Close`) и ценой открытия (`Open`).
- `AdjClose_Low`: Разница между скорректированной ценой закрытия (`Adj Close`) и минимальной ценой (`Low`).
- `AdjClose_High`: Разница между скорректированной ценой закрытия (`Adj Close`) и максимальной ценой (`High`).
- `AdjClose_Close`: Разница между скорректированной ценой закрытия (`Adj Close`) и ценой закрытия (`Close`).

Базовые признаки формируют основу для анализа поведения акций и включают ключевые различия между основными метриками (`Open`, `High`, `Low`, `Close`, `Adj Close`). Например:
- **`Open_High`** позволяет оценить разницу между ценой открытия и максимальной ценой, что может указывать на восходящий тренд.
- **`AdjClose_Low`** помогает учитывать влияние корпоративных действий на минимальную цену в течение дня.

Эти признаки создают дополнительный контекст, который особенно важен для задач временных рядов, где даже небольшие изменения могут быть значимыми для предсказания.

---

**2. Скользящие средние (SMA, EMA) и волатильность**

Выявление трендов и сглаживание данных для получения более информативных признаков. Скользящие средние показывают краткосрочные тренды, а волатильность — уровень риска.

**Признаки**:
  - **SMA** (Simple Moving Average):
    - Примеры: `SMA_3`, `SMA_5`, `SMA_10` — среднее значение за последние 3, 5 и 10 дней.
  - **EMA** (Exponential Moving Average):
    - Похоже на SMA, но с большим весом для недавних данных.
  - **Волатильность** (Volatility):
    - Стандартное отклонение целевой переменной (`Open`) за последние 3, 5 или 10 дней.

---

**3. Процентные изменения**

Анализ относительных изменений объёма торговли и цены. Эти признаки важны для оценки динамики рынка и активности.

**Признаки**:
  - `Volume_Change`: Процентное изменение объёма торговли за день.
  - `Price_Change`: Процентное изменение цены открытия (`Open`) по сравнению с предыдущим днём.
  
---

**4. Признаки тренда и сезонности**

Учёт временных паттернов. Эти признаки позволяют модели учитывать сезонные тренды и их влияние на рынок.

**Признаки**:
  - `Day_of_Week`: День недели (например, понедельник = 0, пятница = 4).
  - `Number_of_Week`: Номер недели года.
  - `Month`: Номер месяца.
  - `Quarter`: Квартал.

---

**5. Лаг-функции**

Предоставление исторического контекста для предсказания. Лаг-признаки важны для анализа временной зависимости, что критично для временных рядов.

**Признаки**:
  - `Lag_1`, `Lag_2`, ..., `Lag_12`: Значения цены открытия (`Open`) за последние 12 дней.

---

**6. Целевая переменная**

Определить метрику, которую нужно предсказывать.

**Признаки**:
  - `Target`: Цена открытия на следующий день (`shift(-1)`).
  - `Low_future`: Минимальная цена на следующий день (`shift(-1)`).

---

**7. Удаление строк с пропусками**

Строки с `NaN`, возникающие из-за вычисления скользящих средних, лагов и других признаков, удаляются для обеспечения корректности данных.

In [8]:
data_review(data)

ПЕРВЫЕ 5 СТРОЧЕК С ДАННЫМИ:


,Open_apple,High_apple,Low_apple,Close_apple,Adj_Close_apple,Volume_apple,Open_microsoft,High_microsoft,Low_microsoft,Close_microsoft,...,Lag_5_apple,Lag_6_apple,Lag_7_apple,Lag_8_apple,Lag_9_apple,Lag_10_apple,Lag_11_apple,Lag_12_apple,Target,Low_future
Date,,,,,,,,,,,,,,,,,,,,,
2019-01-18,39.375000,39.470001,38.994999,39.205002,37.445938,135004000,107.459999,107.900002,105.910004,107.709999,...,38.220001,38.125000,37.822498,37.389999,37.174999,36.132500,35.994999,38.722500,39.102501,38.154999
2019-01-22,39.102501,39.182499,38.154999,38.325001,36.605419,121576000,106.750000,107.099998,104.860001,105.680000,...,37.712502,38.220001,38.125000,37.822498,37.389999,37.174999,36.132500,35.994999,38.537498,37.924999
2019-01-23,38.537498,38.785000,37.924999,38.480000,36.753471,92522400,106.120003,107.040001,105.339996,106.709999,...,37.567501,37.712502,38.220001,38.125000,37.822498,37.389999,37.174999,36.132500,38.527500,37.935001
2019-01-24,38.527500,38.619999,37.935001,38.174999,36.462154,101766000,106.860001,107.000000,105.339996,106.199997,...,38.270000,37.567501,37.712502,38.220001,38.125000,37.822498,37.389999,37.174999,38.869999,38.580002
2019-01-25,38.869999,39.532501,38.580002,39.439999,37.670399,134142000,107.239998,107.879997,106.199997,107.169998,...,38.549999,38.270000,37.567501,37.712502,38.220001,38.125000,37.822498,37.389999,38.947498,38.415001



ОБЩАЯ ИНФОРМАЦИЯ О ДАННЫХ:

<class 'pandas.core.frame.DataFrame'>
DatetimeIndex: 1497 entries, 2019-01-18 to 2024-12-30
Data columns (total 63 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   Open_apple            1497 non-null   float64
 1   High_apple            1497 non-null   float64
 2   Low_apple             1497 non-null   float64
 3   Close_apple           1497 non-null   float64
 4   Adj_Close_apple       1497 non-null   float64
 5   Volume_apple          1497 non-null   int64  
 6   Open_microsoft        1497 non-null   float64
 7   High_microsoft        1497 non-null   float64
 8   Low_microsoft         1497 non-null   float64
 9   Close_microsoft       1497 non-null   float64
 10  Adj_Close_microsoft   1497 non-null   float64
 11  Volume_microsoft      1497 non-null   int64  
 12  Open_nvidia           1497 non-null   float64
 13  High_nvidia           1497 non-null   float64
 14  Low_nvidia            149

,Open_apple,High_apple,Low_apple,Close_apple,Adj_Close_apple,Volume_apple,Open_microsoft,High_microsoft,Low_microsoft,Close_microsoft,...,Lag_5_apple,Lag_6_apple,Lag_7_apple,Lag_8_apple,Lag_9_apple,Lag_10_apple,Lag_11_apple,Lag_12_apple,Target,Low_future
count,1497.000000,1497.000000,1497.000000,1497.000000,1497.000000,1.497000e+03,1497.000000,1497.000000,1497.000000,1497.000000,...,1497.000000,1497.000000,1497.000000,1497.000000,1497.000000,1497.000000,1497.000000,1497.000000,1497.000000,1497.000000
mean,137.650755,139.163856,136.266282,137.797926,136.011285,9.359287e+07,268.117929,270.712625,265.407421,268.188076,...,136.923841,136.783617,136.643552,136.500085,136.357864,136.216343,136.074843,135.935787,137.793083,136.406854
std,53.181516,53.620185,52.778853,53.239493,53.662462,5.170700e+07,94.778261,95.376391,94.042600,94.743296,...,53.049961,53.033484,53.017763,52.995545,52.976210,52.960773,52.945366,52.927489,53.203433,52.799840
min,38.527500,38.619999,37.924999,38.174999,36.462154,2.323470e+07,102.870003,104.099998,102.169998,102.779999,...,37.567501,37.567501,37.567501,37.389999,37.174999,36.132500,35.994999,35.994999,38.527500,37.924999
25%,94.367500,95.980003,93.767502,95.342499,92.903053,5.899130e+07,203.500000,205.039993,200.860001,202.880005,...,93.709999,92.500000,91.962502,91.279999,91.250000,91.102501,91.000000,90.987503,94.839996,93.877502
50%,145.809998,147.520004,144.460007,146.100006,143.956177,8.123540e+07,259.750000,262.299988,256.809998,260.260010,...,145.550003,145.529999,145.490005,145.470001,145.389999,145.350006,145.259995,145.139999,145.820007,144.500000
75%,174.009995,175.880005,172.500000,174.550003,172.605438,1.118507e+08,332.290009,335.200012,328.660004,331.619995,...,173.800003,173.789993,173.750000,173.619995,173.369995,173.330002,173.320007,173.160004,174.029999,172.570007
max,258.190002,260.100006,257.630005,259.019989,259.019989,4.265100e+08,467.000000,468.350006,464.459991,467.559998,...,252.160004,252.160004,252.160004,250.080002,247.990005,247.960007,247.960007,247.960007,258.190002,257.630005


Создание новых признаков позволяет учитывать как краткосрочные тренды и волатильность, так и сезонные закономерности, обеспечивая более полное представление о динамике цен и их взаимосвязях. Это может сделать модель более информативной и эффективной для предсказания цены открытия акций.

### Анализ данных

Вернёмся к общей информации по всем обновлённым данным, которые расположены на пару строк выше, и выполним общий анализ данных. Он направлен на первичное исследование и выявление ключевых характеристик.

---

**1. Общая информация**

Данные охватывают период с **18 января 2019 года** по **30 декабря 2024 года** и включают:
- 1497 записей,
- 63 столбца, большинство из которых представлены в формате `float64`, а также `int32` и `int64`.
- Пропуски и дубликаты отсутствуют, что упрощает их анализ и дальнейшее использование.

Данные включают цены акций (открытия, закрытия, максимальные и минимальные значения), объёмы торгов для нескольких компаний (Apple, Microsoft, Nvidia, Google), а также дополнительные признаки: скользящие средние, волатильность, сезонные показатели и лаги.

---

**2. Статистическая сводка**

Анализ данных показал:
- **Цена открытия (Apple)**:
  - Среднее значение: ~137.65.
  - Диапазон: от **38.53** до **258.19**.
- **Цена закрытия (Apple)**:
  - Среднее значение: ~137.80.
  - Диапазон: от **38.17** до **259.02**.
- **Объём торгов (Apple)**:
  - Среднее значение: ~93.59 млн акций.
  - Максимальный объём: ~426.51 млн акций.

Другие компании, такие как Microsoft, Nvidia и Google, демонстрируют схожие характеристики, включая рост цен к концу временного ряда.

Дополнительные признаки, такие как волатильность, лаги и скользящие средние, обеспечивают возможность учёта трендов и временных зависимостей.

---

**3. Наблюдения**

- **Минимальные значения цен** акций большинства компаний находятся в начале временного ряда, что подтверждает общий тренд роста рынка с 2019 года.
- **Максимальные значения цен** зафиксированы ближе к концу временного ряда, что может быть связано с ростом популярности компаний или улучшением экономических условий.
- Волатильность акций остаётся значительной, что может указывать на активные колебания цен в течение всего периода.

---

**4. Выводы**

- Данные готовы к дальнейшему анализу, включая построение моделей и визуализацию.
- Признаки, связанные с трендами (скользящие средние), сезонностью (день недели, месяц, квартал) и историческими значениями (лаги), помогут улучшить предсказательную способность модели.
- Следует изучить корреляцию между компаниями для выявления общих рыночных закономерностей и их влияния на поведение акций.



## Обучение
### Functions

In [9]:
def optimize_df(df):
    """Уменьшение размерности датафрейма"""
    initial_memory = df.memory_usage(deep=True).sum()
    
    for col in df.columns:
        col_type = df[col].dtype
        
        if col_type == 'float64' or col_type == 'float32':
            # Проверяем, можно ли привести float к int
            if df[col].apply(float.is_integer).all():
                df[col] = pd.to_numeric(df[col], downcast='integer')
            else:
                df[col] = pd.to_numeric(df[col], downcast='float')
        
        elif col_type == 'int64' or col_type == 'int32':
            df[col] = pd.to_numeric(df[col], downcast='integer')
        
        elif col_type == 'object':
            num_unique_values = len(df[col].unique())
            num_total_values = len(df[col])
            if num_unique_values / num_total_values < 0.5:
                df[col] = df[col].astype('category')
        
        elif col_type == 'bool':
            df[col] = df[col].astype('int8')
    
    final_memory = df.memory_usage(deep=True).sum()
    memory_reduction = (initial_memory - final_memory) / initial_memory * 100
    
    print(f"Память до оптимизации: {initial_memory} байт")
    print(f"Память после оптимизации: {final_memory} байт")
    print(f"Сокращение памяти: {memory_reduction:.2f}%")
    
    return df

In [10]:
def plot_feature_importances(model, feature_names, model_name):
    """
    Функция для вывода важности признаков для модели.

    Parameters:
    model (sklearn model): Обученная модель.
    feature_names (list): Список названий признаков.
    model_name (str): Название модели.
    """
    if hasattr(model, 'feature_importances_'):
        importances = model.feature_importances_
    elif hasattr(model, 'coef_'):
        importances = model.coef_
    else:
        raise ValueError(f"Model {model_name} does not have feature importances or coefficients.")

    # Создание DataFrame для удобного отображения
    feature_importance_df = pd.DataFrame({
        'Feature': feature_names,
        'Importance': importances
    })
    feature_importance_df = feature_importance_df.sort_values(by='Importance', ascending=False)
    
    # Построение графика
    plt.figure(figsize=(10, 6))
    sns.barplot(x='Importance', y='Feature', data=feature_importance_df)
    plt.title(f'Feature Importances for best model - {model_name}')
    plt.show()

### Подготовка к обучению

### Train

## Выбор лучшей модели

## Выводы